### PR TITLE
Add microplugin@0.0.3 w/ npm auto-update

### PR DESCRIPTION
**Pull request by mercyg**
Pull request for issue: #
Related issue(s): # #

Checklist for **Pull request** or **lib adding request issue** follows the conventions.

Note that if you are using a distribution purpose repository/package, please also provide the url and other related info like popularity of the source code repo/package.

# Profile of the lib
 * Git repository (required):https://github.com/brianreavis/microplugin.js
 * Official website (optional, not the repository):
 * NPM package url (optional):https://www.npmjs.com/package/microplugin
 * License and its reference:
 * GitHub / Bitbucket popularity (required):
   - Count of watchers:
   - Count of stars:
   - Count of forks:
 * NPM download stats (optional):
   - Downloads in the last day:
   - Downloads in the last week:
   - Downloads in the last month:

# Essential checklist
 * [ ] I'm the author of this library
   * [ ] I would like to add link to the page of this library on CDNJS on website / readme
 * [x] This lib was not found on cdnjs repo
 * [x] No already exist / duplicated issue and PR
 * [x] The lib has notable popularity
   * [x] More than 100 [Stars / Watchers / Forks] on [GitHub / Bitbucket]
   * [x] More than 500 downloads stats per month on npm registry
 * [x] Project has public repository on famous online hosting platform (or been hosted on npm)

# Auto-update checklist
 * [x] Has valid tags for each versions (for git auto-update)
 * [x] Auto-update setup
 * [x] Auto-update target/source is valid.
 * [x] Auto-update filemap is correct.

# Git commit checklist
 * [x] The first line of commit message is less then 50 chars, be clean and clear, easy to understand.
 * [ ] The parent of the commit(s) in the PR is not old than 3 days.
 * [x] Pull request is sending from a non-master branch with meaningful name.
 * [x] Separate unrelated changes into different commits.
 * [ ] Use rebase to squash/fixup dummy/unnecessary commits into only one commit.
 * [x] Close corresponding issue in commit message
 * [x] Mention related issue(s), people in commit message, comment.

### CODE REVIEW

**Comment by maruilian11** [comment]
hi @mercyg, please fulfill the lib info in comment, which will help us review your PR more quickly.

**Comment by maruilian11** [comment]
related issue https://github.com/cdnjs/cdnjs/issues/9973

**Comment by maruilian11** [comment]
@mercyg there are two ways to add this lib:
1. add package.json with `version` field and 0.0.3 real files(included minified files). 
2. only package.json without `version` field.

which one do you prefer? I will help you in the way you like.

**Comment by mercyg** [comment]
Hi @maruilian11 I would like to add package.json  with version field.

File: ajax/libs/microplugin/package.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "microplugin",
```
Comment (maruilian11) [inline ajax/libs/microplugin/package.json]: please add `"filename": "microplugin.min.js",`

File: ajax/libs/microplugin/package.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    "extensibility",
+    "plugins",
+    "architecture",
+    "require",
+    "dependencies"
+  ],
+  "description": "A lightweight plugin / dependency system for javascript libraries.",
+  "version": "0.0.3",
+  "author": "Brian Reavis <brian@thirdroute.com>",
+  "main": "src/microplugin.js",
```
Comment (maruilian11) [inline ajax/libs/microplugin/package.json]: please remove `main`

File: ajax/libs/microplugin/package.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git://github.com/brianreavis/microplugin.js.git"
+  },
+  "dependencies": {},
+  "engines": {
```
Comment (maruilian11) [inline ajax/libs/microplugin/package.json]: please remove `dependencies` and `engines`

File: ajax/libs/microplugin/package.json [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  "npmName": "microplugin",
+  "npmFileMap": [
+    {
+      "basePath": "/src",
+      "files": [
+        "**/*"
+      ]
+    }
+  ],
+  "filename": "/microplugin.js"
```
Comment (maruilian11) [inline ajax/libs/microplugin/package.json]: we prefer put `filename` in the previous field of package.json, thanks

File: ajax/libs/microplugin/package.json [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    "url": "git://github.com/brianreavis/microplugin.js.git"
+  },
+  "dependencies": {},
+  "engines": {
+    "node": "*"
+  },
+  "npmName": "microplugin",
+  "npmFileMap": [
+    {
+      "basePath": "/src",
```
Comment (maruilian11) [inline ajax/libs/microplugin/package.json]: please just use `src` here

File: ajax/libs/microplugin/package.json [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      "files": [
+        "**/*"
```
Comment (maruilian11) [inline ajax/libs/microplugin/package.json]: we only use "**/*" in `dist` or `build` folder, please use `microplugin*` here, thanks

**Comment by mercyg** [comment]
Thanks for the review,  I will try to make the correction. Sorry for all the error, I am a newbie for open source project:)

**Comment by maruilian11** [comment]
@mercyg It's ok. Take your time 👍

**Comment by mercyg** [comment]
@maruilian11 do I have to close this pull request in order to make the change and make another pull request?

**Comment by mercyg** [comment]
@maruilian11  I have made the correction please try to look over my pull request

File: ajax/libs/microplugin/package.json [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  "npmName": "microplugin",
+  "npmFileMap": [
+    {
+      "basePath": "src",
+      "files": [
+        "microplugin*"
+      ]
+    }
+  ]
+
```
Comment (maruilian11) [inline ajax/libs/microplugin/package.json]: please remove this blank line, thanks.

File: ajax/libs/microplugin/package.json [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    "require",
+    "dependencies"
+  ],
+  "description": "A lightweight plugin / dependency system for javascript libraries.",
+  "version": "0.0.3",
+  "author": "Brian Reavis <brian@thirdroute.com>",
+  "repository": {
+    "type": "git",
+    "url": "git://github.com/brianreavis/microplugin.js.git"
+  },
```
Comment (maruilian11) [inline ajax/libs/microplugin/package.json]: please also add `"license": "Apache-2.0",`, thanks.

**Comment by mercyg** [comment]
@maruilian11 I have removed the blank line please take a look over my pull request

**Comment by mercyg** [comment]
@maruilian11 added the license

**Comment by maruilian11** [comment]
hi @mercyg, could you please tell me what minify tool you used? thanks!

**Comment by mercyg** [comment]
@maruilian11 I used an online minifying tool and added the file locally

**Comment by maruilian11** [comment]
after reviewing, the all commits need to be squashed, would you like to squash them? or I can do it for you. thanks!

**Comment by maruilian11** [comment]
@mercyg could you provide the online minify tool url? thanks!

**Comment by mercyg** [comment]
@maruilian11 https://jscompress.com/

**Comment by mercyg** [comment]
@maruilian11 I don't know how to squash commit

**Comment by maruilian11** [comment]
@mercyg do you use command line? or GitHub website UI?

**Comment by mercyg** [comment]
@maruilian11 I prefer command line

**Comment by PeterDaveHello** [comment]
@mercyg If you are using command line, it won't be too hard to squash the commits, we can help you if you'd like to learn how to do it.

**Comment by mercyg** [comment]
@PeterDaveHello sounds great I would like to learn how to do it

**Comment by mercyg** [comment]
@maruilian11 $ git rebase -i HEAD~4 is this the command to squash the four commits

**Comment by maruilian11** [comment]
@mercyg yes! But I think we should squash these 5 commit.

**Comment by mercyg** [comment]
@maruilian11 Ok I will try that

**Comment by mercyg** [comment]
@maruilian11 Cannot rebase: You have unstaged changes.
Please commit or stash them

I am getting this error

**Comment by PeterDaveHello** [comment]
@mercyg You need to make sure your working directory is clean, at least there is no tracked files been modified, check the result of git status

**Comment by mercyg** [comment]
@PeterDaveHello I did try checking for git status and I try commit it but I am still getting the same error

**Comment by PeterDaveHello** [comment]
@mercyg What's the the final git status result? There must be something modified.

**Comment by mercyg** [comment]
@PeterDaveHello   new files yourself (see 'git help status').
no changes added to commit (use "git add" and/or "git commit -a")

**Comment by PeterDaveHello** [comment]
@mercyg Mind to give us a screnshot or something like that?

On my computer, the clean repository will show git status result as:
On branch master
nothing to commit, working tree clean

If there is a modified file, it'll show something like this:
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   package.json

no changes added to commit (use "git add" and/or "git commit -a")

**Comment by mercyg** [comment]
screen shot 2017-02-14 at 2 12 41 am

**Comment by mercyg** [comment]
@PeterDaveHello

**Comment by PeterDaveHello** [comment]
@mercyg so obviously there are many many files there been touched that you need to reset them first, try \`git reset —hard`

**Comment by mercyg** [comment]
@PeterDaveHello ok I have done that should I do the git rebase now

**Comment by PeterDaveHello** [comment]
@mercyg if it's now clean, you'll be able to squash now

**Comment by mercyg** [comment]
@PeterDaveHello I am getting the same error again

**Comment by PeterDaveHello** [comment]
@mercyg So it's fixed but broken again? Or it's not fixed at all? One another question, have you enabled sparseCheckout?

**Comment by mercyg** [comment]
@PeterDaveHello how do you enable that ?

**Comment by PeterDaveHello** [comment]
@mercyg https://github.com/cdnjs/cdnjs#important

**Comment by PeterDaveHello** [comment]
@mercyg https://github.com/cdnjs/cdnjs/blob/master/.github/CONTRIBUTING.md#a-temporary-version-of-the-contributing-doc

**Comment by PeterDaveHello** [comment]
@mercyg Take a look at the reference in readme/contributing doc please.

**Comment by mercyg** [comment]
@PeterDaveHello yes I have enabled sparseCheckout

**Comment by PeterDaveHello** [comment]
@mercyg can you try git reset --hard again, and see if the not used library not exist under ajax/libs/ anymore?

**Comment by mercyg** [comment]
@PeterDaveHello the libraries still exist after hard rest

**Comment by PeterDaveHello** [comment]
@mercyg All of them? Then maybe you didn't enabled it with success, can you make sure "git config core.sparseCheckout" outputs true? (C not c of Checkout)

**Comment by mercyg** [comment]
@PeterDaveHello Yes it returns true

**Comment by PeterDaveHello** [comment]
@mercyg hmmmm .... that's odd, so the reset never fixed anything? Or it's just broken again?

**Comment by mercyg** [comment]
@PeterDaveHello it doesn't look like it fixed anything

**Comment by PeterDaveHello** [comment]
@mercyg can you show me your .git/info/sparse-checkout? Though I think it's now becoming two problems - 1. your reset hard not work, 2. your sparseCheckout not work, hmmmm

**Comment by mercyg** [comment]
@PeterDaveHello Do you want me to run .git/info/sparse-checkout

**Comment by PeterDaveHello** [comment]
@mercyg nope, that's the config file, looks like you missed it ... or you'll know what's the file for, maybe you can check the doc again?

**Comment by mercyg** [comment]
@PeterDaveHello ohh ok my bad I will check it again

**Comment by mercyg** [comment]
@PeterDaveHello I don't know what I have to check out for this issue because in the instruction it says "check set what you want to checkout only, write to .git/info/sparse-checkout".  So should I just checkout the jquery as in like the example?

**Comment by PeterDaveHello** [comment]
@mercyg you should put ajax/libs/microplugin in it as you are working on this lib.

**Comment by mercyg** [comment]
@PeterDaveHello git hard reset worked:)

**Comment by PeterDaveHello** [comment]
@mercyg Nice, let's keep going!

**Comment by mercyg** [comment]
@PeterDaveHello  When I am run  git rebase -i HEAD~5 I got into the vim editor. I don't know what command that I have to use?

**Comment by mercyg** [comment]
@PeterDaveHello I have successfully rebase it locally so now do I push it to Github?

**Comment by PeterDaveHello** [comment]
@mercyg You'll also need to update your master branch, and rebase your branch with the latest master branch :)

**Comment by mercyg** [comment]
@PeterDaveHello how do I do that?

**Comment by PeterDaveHello** [comment]
@mercyg pull master to the latest, then "git rebase master mybranch"

**Comment by mercyg** [comment]
@PeterDaveHello so is it merging my master branch with microplugin branch?

**Comment by PeterDaveHello** [comment]
@mercyg nope, it's rebase, not the same

**Comment by mercyg** [comment]
@PeterDaveHello I am sorry for my silly question but how do you pull master to the latest?

**Comment by PeterDaveHello** [comment]
@mercyg no worries, you should add cdnjs as one of your git remotes, try:
git remote add cdnjsUpstream https://github.com/cdnjs/cdnjs

then you can do: git pull cdnjsUpstream master

**Comment by mercyg** [comment]
@PeterDaveHello after rebasing the master branch with my branch I got this "Current branch importer-microplugin-0.0.3 is up to date."

**Comment by PeterDaveHello** [comment]
@mercyg which means you master branch is not up to date, so should update master branch first

**Comment by mercyg** [comment]
@PeterDaveHello when I make the pull request I got this error at the end "error: Sparse checkout leaves no entry on working directory"

**Comment by PeterDaveHello** [comment]
@mercyg can you provide what's the exactly command you input and the output you got? The info need to be in more detail and precise then I can help solve the problem, thanks.

**Comment by maruilian11** [comment]
hi @mercyg, could you try
1. `git checkout master`
2. `git pull https://github.com/cdnjs/cdnjs.git master` // you can check if the latest commit is same as our latest commit(on master) on website
3. `git checkout importer-microplugin-0.0.3`
4. `git rebase master`// after it done, the latest commit on our master should behind your commit
5. then you can push to this branch

**Comment by mercyg** [comment]
@maruilian11 when I do the second instruction "git pull https://github.com/cdnjs/cdnjs.git master"  I get this error "error: Sparse checkout leaves no entry on working directory"

**Comment by maruilian11** [comment]
@mercyg could you show us your sparse-checkout content? (it is usually put in `.git/info/`)

**Comment by PeterDaveHello** [comment]
@mercyg you should also add the other sparseCheckout template content in the config, then the problem will be fixed I think.

**Comment by mercyg** [comment]
@maruilian11 sparesCheckout contain "/ajax/libs/microplugin"

**Comment by mercyg** [comment]
@PeterDaveHello what is the other sparseCheckout template?

**Comment by maruilian11** [comment]
https://github.com/cdnjs/cdnjs/blob/master/tools/sparse-checkout.template
@mercyg does your `sparse-checkout` look like:
```
/build
/documents
/test
/tools
/MIT-LICENSE
/hooks
/MAINTAINERS
/*.yml
/*.md
/*.json
/*.sh
/*.js
/ajax/libs/microplugin/*
```

**Comment by PeterDaveHello** [comment]
@mercyg just as the example in the doc, or use https://github.com/cdnjs/cdnjs/blob/master/tools/sparse-checkout.template

**Comment by PeterDaveHello** [comment]
@mercyg here is a trick that you don't need to checkout back to the master branch but still update it:

`git fetch https://github.com/cdnjs/cdnjs +master:master`

Then you can do the rebase!

**Comment by mercyg** [comment]
@maruilian11 the git pull worked the problem was spares-checkout file was only /ajax/libs/microplugin/* and I updated it with what it should be.

**Comment by PeterDaveHello** [comment]
@mercyg once you updated it, you should be able to checkout back to master branch, the reason why you failed is because your config didn't mention any file exist in master branch, so git didn't know what should it do.

**Comment by mercyg** [comment]
@PeterDaveHello  after I did the "git rebase master" and then "git checkout importer-microplugin-0.0.3" I get this error that says your branch have diverged.

**Comment by PeterDaveHello** [comment]
@mercyg where did you do the rebase action? On `importer-microplugin-0.0.3`?

**Comment by mercyg** [comment]
@PeterDaveHello  yes

**Comment by mercyg** [comment]
@PeterDaveHello this is the error

Your branch and 'origin/importer-microplugin-0.0.3' have diverged,
and have 296 and 5 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

**Comment by PeterDaveHello** [comment]
@mercyg this is not a error, just a warning message, I think you did something not correct, so the update and rebase didn't really work.

Please make sure:

1. You pull the master branch to the latest
2. Then do  `git rebase master` on `importer-microplugin-0.0.3` branch, or directly do`git rebase master importer-microplugin-0.0.3` in a single step

Then everything should be good.

**Comment by mercyg** [comment]
@PeterDaveHello I did that correctly. What if I did pull to merge the remote into the branch just like in the warning or can I force push from my branch?

**Comment by PeterDaveHello** [comment]
@mercyg There shouldn't be any merge commit by your side, if there is, try to pull with '--rebase', if there is no merge commit from your side, just push with '--force' to update the branch.

**Comment by mercyg** [comment]
@PeterDaveHello I did the git push --force.

**Comment by PeterDaveHello** [comment]
@mercyg Nice, we'll see if it passed the test. (If it didn't, you can also try to fix it by reading the error!)

**Comment by PeterDaveHello** [comment]
@mercyg Okay, looks you passed the test!

**Comment by mercyg** [comment]
@PeterDaveHello great :) Thank you very much for all the help:)

**Comment by PeterDaveHello** [comment]
Glad that we solved almost all the issues here, just the last thing, let's write a better commit message, shall we?

It'll be better to use commit message like this:
```
Add microplugin@0.0.3 with git auto-update

Close #9973, cc @brianreavis
```

So that the issue will be closed, the author will know that the lib is now on CDNJS, and we know that you added auto-update config.

Please note that the second line of the commit message should always be empty, if you are interested in writing good commit message, this is a reference for you: https://chris.beams.io/posts/git-commit/

Let's fix the last thing and get it merged!

**Comment by mercyg** [comment]
@PeterDaveHello Can I make the commit message change in GitHub or do I have to do it on command line?

**Comment by PeterDaveHello** [comment]
@mercyg Need to be under command, and use rebase to reword the commit, very like how you squash them.

**Comment by mercyg** [comment]
@PeterDaveHello I have updated the commit message :)

**Comment by PeterDaveHello** [comment]
@mercyg thanks. @pvnr0082t please also help review this one, thanks.

**Comment by mercyg** [comment]
@pvnr0082t what id did I tag?

**Comment by pvnr0082t** [comment]
@mercyg You tagged br**ai**nreavis, but the author name is br**ia**nreavis.

**Comment by mercyg** [comment]
@pvnr0082t Oh ok I will change that

**Comment by mercyg** [comment]
@pvnr0082t I have updated the commit message

**Comment by PeterDaveHello** [comment]
@mercyg here we go: https://cdnjs.com/libraries/microplugin, thanks for your contribution!

**Comment by mercyg** [comment]
@PeterDaveHello thank you for the help. I have learned a lot:)

**Comment by PeterDaveHello** [comment]
@mercyg that's nice, wanna learn more? Haha!

**Comment by mercyg** [comment]
@PeterDaveHello yeah definitely.

**Comment by PeterDaveHello** [comment]
@mercyg glad you like it, maybe you can start from picking other issues.

**Comment by mercyg** [comment]
@PeterDaveHello Yeah, I will do that

**Pull request merged**